p2p/peer_behaviour.go

```diff
@@ -0,0 +1,137 @@
+package p2p
+
+import (
+	"errors"
+	"sync"
+)
+
+// Type of error behaviour a peer can perform.
+type ErrorPeerBehaviour int
+
+const (
+	ErrorPeerBehaviourUnknown = iota
+	ErrorPeerBehaviourBadMessage
+	ErrorPeerBehaviourMessageOutofOrder
+)
+
+// Type of good behaviour a peer can perform.
+type GoodPeerBehaviour int
+
+const (
+	GoodPeerBehaviourVote = iota + 100
+	GoodPeerBehaviourBlockPart
+)
+
+// PeerBehaviour provides an interface for reactors to signal the behaviour
+// of peers synchronously to other components.
+type PeerBehaviour interface {
+	Behaved(peerID ID, reason GoodPeerBehaviour) error
+	Errored(peerID ID, reason ErrorPeerBehaviour) error
+}
+
+type SwitchPeerBehaviour struct {
+	sw *Switch
+}
+
+// Reports the ErrorPeerBehaviour of peer identified by peerID to the Switch.
+func (spb *SwitchPeerBehaviour) Errored(peerID ID, reason ErrorPeerBehaviour) error {
+	peer := spb.sw.Peers().Get(peerID)
+	if peer == nil {
+		return errors.New("Peer not found")
+	}
+
+	spb.sw.StopPeerForError(peer, reason)
+	return nil
+}
+
+// Reports the GoodPeerBehaviour of peer identified by peerID to the Switch.
+func (spb *SwitchPeerBehaviour) Behaved(peerID ID, reason GoodPeerBehaviour) error {
+	peer := spb.sw.Peers().Get(peerID)
+	if peer == nil {
+		return errors.New("Peer not found")
+	}
+
+	spb.sw.MarkPeerAsGood(peer)
+	return nil
+}
+
+// Return a new switchPeerBehaviour instance which wraps the Switch.
+func NewSwitchPeerBehaviour(sw *Switch) *SwitchPeerBehaviour {
+	return &SwitchPeerBehaviour{
+		sw: sw,
+	}
+}
+
+// storedPeerBehaviour serves a mock concrete implementation of the
+// PeerBehaviour interface used in reactor tests to ensure reactors
+// produce the correct signals in manufactured scenarios.
+type StoredPeerBehaviour struct {
+	eb  map[ID][]ErrorPeerBehaviour
+	gb  map[ID][]GoodPeerBehaviour
+	mtx sync.RWMutex
+}
+
+// GettablePeerBehaviour provides an interface for accessing ErrorPeerBehaviours
+// and GoodPeerBehaviour recorded by an implementation of PeerBehaviour
+type GettablePeerBehaviour interface {
```

Comment:
By now it seems like this interface is superfluous. The comment already reveals a tight coupling to PeerBehaviour, but really the only place we going to need this is for `StoredPeerBehaviour`.  Why don't we dissolve this interface entirely and leave those two methods just as sturct methods on `StoredPeerBehaviour`, if the caller wants to call them they need to be aware they are handling that struct, esp. now that's public this will work fine. Should we end up with more concrete implementations exposing the same methods we can reassess the need for such an abstraction.